docs/rfcs/099-safekeeper-s3-integration.md

```diff
@@ -0,0 +1,86 @@
+# What
+
+Safe keeper needs to backup WAL to S3
+
+S3 is used across the Neon for a variety of components and sharing code will reduce code maintanance cost 
+
+# Why
+
+In situations when other sources of data are compromised transaction log (WAL) produced by compute becomes the ultimate recovery solution 
+
+see https://github.com/zenithdb/zenith/issues/1169 
+
+
+# Proposal
+
+
+## A common S3 integration
+
+Given that S3 integration is common between both Safekeepers and Page Servers a common library can be built
+The common requirement is an ability for asynchrounous upload of objects to a Remote Storage.
+
+Objects can be represented by a byte stream or by a file, optional compression can be applied on either
+Remote storage shuold have production implementation and test implementations. Production impementation is built to work with S3 API compatible blob storage while test implementation is based on a filesystem
+
+ Commmon operations to be supported:
+  1. Upload object (inputs: an object and a destination)
+  2. List objects (inputs: an object filter)  
+  *this is irrelevant to Safekeepers and no discussed here*
+  3. Download object (inputs: a destination)  
+  *this is irrelevant to Safekeepers and no discussed here*
+
+In order to isolate tenants interface when created need to require a timeline.
+This should minimize chances of object clobbering by others.
+
+## Safe keeper specific
+
+**WAL Backup** is a layer above basic s3 integration should allow uploading objects and perform maintain **WAL Backup LSN** metadata, it's a glue between "Storage" object and S3 infra.
+
+Storage object would provide a notification interface to which WAL Backup would subscribe. Notifications are fired when segment is full (e.g. no more data fits into the segment and ".partial" suffix is removed).
+WAL Backup will subscribe to the notification and will upload a segment, upon completion it will attempt to advance WAL Backup LSN in etcd.
```

Comment:
We also should wait for commit_lsn to be advanced beyond segment. Otherwise bytes can be modified after upload.

Reply:
given that we're writing sequentially in what situations would the data in the complete file get modified?

Reply:
Yeah, we are writing sequentially **most** of the time, but with consensus there are cases when we will jump back to the lower LSN for the writes. In every election LSN is decided based on the responses from the quorum of safekeepers, and there is a possibility to write something to a single safekeeper in a term `T`, but get a quorum of responses from the other ones in the next term `T+1`. This is exactly the case when WAL will be truncated to a lower LSN, when the election is finished in the term `T+1`.

Reply:
👍 Agree I'll amend the diagram. if commit LSN is the highest LSN for which all previous log records were persisted, then it's fine (I was under impression that we had Commit LSN as the highest LSN we achieved a quorum write)